### PR TITLE
[2.2.0] Fix deadlock when releasing segments and removing nodes concurrently

### PR DESCRIPTION
**Pull request by yah01**
/kind bug
related #22575 #22360

### CODE REVIEW

File: internal/querynode/distribution.go [hunk 1]
```diff
@@ -208,10 +207,11 @@ func (d *distribution) RemoveDistributions(releaseFn func(), sealedSegments ...S
...
 		}
 	}
-
 	d.offlines.Add(delta)
-
 	<-d.genSnapshot()
+
+	d.mut.Unlock()
+
 	releaseFn()
```
Comment (congqixia) [inline internal/querynode/distribution.go:215]: `releaseFn` shall be executed before lock released

File: internal/querynode/shard_cluster.go [hunk 2]
```diff
@@ -449,10 +449,14 @@ func (sc *ShardCluster) watchSegments(evtCh <-chan segmentEvent) {
...
-// getNode returns shallow copy of shardNode
-func (sc *ShardCluster) getNode(nodeID int64) (*shardNode, bool) {
+// GetNode returns shallow copy of shardNode
+func (sc *ShardCluster) GetNode(nodeID int64) (*shardNode, bool) {
 	sc.mut.RLock()
 	defer sc.mut.RUnlock()
+	return sc.getNode(nodeID)
+}
+
+func (sc *ShardCluster) getNode(nodeID int64) (*shardNode, bool) {
```
Comment (congqixia) [inline internal/querynode/shard_cluster.go:459]: Why change get node to public definition? Maybe `getNode` and `getNodeInternal` (forgive my naming sense). The point is this function shall be private

File: internal/querynode/distribution.go (hunk 1 above)
Comment (yah01) [reply internal/querynode/distribution.go:215]: Why? I didn't see any distribution related code in release fn

Comment (congqixia) [reply internal/querynode/distribution.go:215]: to prevent incoming search reading any possible growing segments need to be released

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**